### PR TITLE
[gce] GCE Binding initial contribution

### PR DESCRIPTION
**Pull request by clinique**
This is a rewrite of the IPX binding for OH1. Renamed to GCE, the producer of the equipments because we may implement more devices in the future.
Seems to be working fine right now. Documentation as to be done.

### CODE REVIEW

**Comment by wborn** [comment]
Initially I thought this PR was for a [Google Compute Engine](https://en.wikipedia.org/wiki/Google_Compute_Engine) Binding. But apparently it's for something completely else. I've probably been using the GCE too much. :wink:

**Comment by Hilbrand** [comment]
Maybe you've already seen it. But you miss the binding in the `bundle/pom.xml` and the `feature.xml` is missing.

**Comment by clinique** [comment]
@Hilbrand  : maybe, but I also see I did put a big mess in my last commit :-)

**Comment by Hilbrand** [comment]
Not necessarily. I think it was caused because the pr was originally made against master, and that caused a big diff when you pushed your rebased changes. So I changed it to 2.5.x. and its ok again.

**Comment by clinique** [comment]
> Not necessarily. I think it was caused because the pr was originally made against master, and that caused a big diff when you pushed your rebased changes. So I changed it to 2.5.x. and its ok again.

Thanks, nice to you, but in the meantime I recreated the PR.

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800v3Handler.java [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+            return;
+        }
+        logger.info("Can not handle command '{}' on channel '{}'", command, channelUID);
+    }
+
+    @Override
+    public void channelLinked(ChannelUID channelUID) {
+        logger.debug("channelLinked: {}", channelUID);
+        final String channelId = channelUID.getId();
+        if (channelUID.getIdWithoutGroup().chars().allMatch(Character::isDigit)) {
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800v3Handler.java:243]: Maybe put this check in a method with the channelUID as parameter. The check is done in multiple places. It looks like you could also simply check if it contains `-` or not or 
```suggestion
        if (!channelUID.getIdWithoutGroup().contains("-")) {
```

**Review by Hilbrand — CHANGES_REQUESTED** [review]
I've done a review. I skipped readme as it needs to be done as you mentioned. You also need to update the CODEOWNERS and `bom/openhab-addons/pom.xml`.

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800v3Handler.java [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+        }
+    }
+
+    @Override
+    public void channelUnlinked(ChannelUID channelUID) {
+        super.channelUnlinked(channelUID);
+        portDatas.remove(channelUID.getId());
+    }
+
+    public void resetCounter(int counter) {
```
[the remark this answers is not in the document]
Comment (clinique) [reply bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800v3Handler.java:265]: They are used by IPX Actions

**Comment by Hilbrand** [comment]
ps you need to sign off with your real name.

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/config/AnalogInputConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * The {@link AnalogInputConfiguration} class holds configuration informations of
+ * an ipx800 Analog Input port.
+ *
+ * @author Gaël L'hopital - Initial contribution
+ */
+@NonNullByDefault
+public class AnalogInputConfiguration extends CounterConfiguration {
+    public long histeresis;
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/config/AnalogInputConfiguration.java:25]:
```suggestion
    public long hysteresis;
```

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+ * reading, writing and disconnecting from the Ipx800.
+ *
+ * @author Seebag - Initial contribution on OH1
+ * @author Gaël L'hopital - Ported and adapted for OH2
+ */
+@NonNullByDefault
+public class Ipx800DeviceConnector extends Thread {
+    private final Logger logger = LoggerFactory.getLogger(Ipx800DeviceConnector.class);
+    private static final int DEFAULT_SOCKET_TIMEOUT = 5000;
+    private static final int DEFAULT_RECONNECT_TIMEOUT = 5000;
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java:38]: It's good practice to append the unit to the field name e.g. TIMEOUT_SEC.

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    }
+
+    /**
+     * Disconnect the device
+     */
+    private void disconnect() {
+        if (connected) {
+            logger.debug("Disconnecting");
+            try {
+                client.close();
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java:91]: I think you have to close the readers and writers *you* created here.

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        }
+        out.println("GetIn01");
+        out.flush();
+        waitingKeepaliveResponse = true;
+    }
+
+    @Override
+    public void run() {
+        interrupted = false;
+        while (!interrupted) {
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java:128]: Why not use Thread's flag?
```suggestion
        while (!interrupted()) {
```

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+                failedKeepalive = 0;
+                connect();
+                while (!interrupted) {
+                    if (failedKeepalive > MAX_KEEPALIVE_FAILURE) {
+                        throw new IOException("Max keep alive attempts has been reached");
+                    }
+                    try {
+                        String command = in.readLine();
+                        waitingKeepaliveResponse = false;
+                        parser.ifPresent(parser -> parser.unsollicitedUpdate(command));
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java:140]:
```suggestion
                        parser.ifPresent(parser -> parser.unsolicitedUpdate(command));
```

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+            }
+        }
+    }
+
+    private void handleException(Exception e) {
+        if (e instanceof SocketTimeoutException) {
+            sendKeepalive();
+            return;
+        } else if (e instanceof IOException) {
+            logger.info("Communication error : '{}', will retry in {} ms", e, DEFAULT_RECONNECT_TIMEOUT);
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java:163]: Logging to info.

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800EventListener.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+@NonNullByDefault
+public interface Ipx800EventListener {
+
+    /**
+     * Procedure for receive data from IPX800 controller.
+     *
+     * @param port Port (kind and number) receiving update
+     * @param value value updated
+     */
+    void dataReceived(String port, Double value);
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800EventListener.java:31]:
```suggestion
    void dataReceived(String port, double value);
```

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800MessageParser.java [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+     *
+     * @param data
+     */
+    public void unsollicitedUpdate(String data) {
+        if (IO_PATTERN.matcher(data).matches()) {
+            String portKind = "GetOutputs".equalsIgnoreCase(expectedResponse) ? RELAY_OUTPUT
+                    : "GetInputs".equalsIgnoreCase(expectedResponse) ? DIGITAL_INPUT : null;
+            if (portKind != null) {
+                for (int count = 0; count < data.length(); count++) {
+                    setStatus(portKind + String.valueOf(count + 1), new Double(data.charAt(count) - '0'));
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800MessageParser.java:76]: Same for below.
```suggestion
                    setStatus(portKind + (count + 1), (double) data.charAt(count) - '0');
```

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800MessageParser.java [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+                    case COUNTER:
+                        portNumShift = -1; // Align counters on 1 based array
+                    case ANALOG_INPUT: {
+                        int portNumber = Integer.parseInt(statusPart[0].substring(1));
+                        setStatus(portKind + String.valueOf(portNumber + portNumShift + 1),
+                                Double.parseDouble(statusPart[1]));
+                    }
+                }
+            }
+        } else if (!"".equals(expectedResponse)) {
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800MessageParser.java:102]:
```suggestion
        } else if (!expectedResponse.isEmpty()) {
```

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800MessageParser.java [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+            setStatus(expectedResponse, Double.parseDouble(data));
+        }
+
+        expectedResponse = "";
+    }
+
+    private void setStatus(String port, Double value) {
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800MessageParser.java:109]:
```suggestion
    private void setStatus(String port, double value) {
```

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800MessageParser.java [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    public void errorOccurred(Exception e) {
+        logger.warn("Error received from connector : {}", e.getMessage());
+        listener.ifPresent(l -> l.errorOccurred(e));
+    }
+
+    public synchronized void getValue(String channelId) {
+        logger.debug("Requested value for {}", channelId);
+        if ("".equals(expectedResponse)) { // Do not send a request is something is expected
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800MessageParser.java:148]: See above

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800v3Handler.java [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+    }
+
+    public Ipx800v3Handler(Thing thing) {
+        super(thing);
+        logger.debug("Create a IPX800 Handler for thing '{}'", getThing().getUID());
+    }
+
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.OFFLINE);
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800v3Handler.java:96]: After the thing status is set to OFFLINE, the framework starts to work with this Thing: `handleCommand()` and `dispose()` can be invoked (concurrently). You might want to remove it as you set the state at the end of the method.

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800v3Handler.java [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+
+        logger.debug("Initializing IPX800 handler for uid '{}'", getThing().getUID());
+
+        connector = new Ipx800DeviceConnector(configuration.hostname, configuration.portNumber);
+        connector.setName("OH-binding-" + getThing().getUID());
+        connector.setDaemon(true);
+
+        parser = Optional.of(new Ipx800MessageParser(connector, this));
+
+        updateStatus(ThingStatus.ONLINE);
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800v3Handler.java:108]: If I see correctly the connection is not yet established here. Would UNKNOWN suit better?

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800v3Handler.java [hunk 16]
```diff
@@ -0,0 +1,282 @@
...
+
+        Channel channel = thing.getChannel(channelUID.getId());
+        if (isValidPortId(channelUID) && RELAY_OUTPUT.equalsIgnoreCase(channelUID.getGroupId())
+                && command instanceof OnOffType && channel != null) {
+            RelayOutputConfiguration config = channel.getConfiguration().as(RelayOutputConfiguration.class);
+            parser.ifPresent(p -> p.setOutput(channelUID.getIdWithoutGroup(),
+                    (OnOffType) command == OnOffType.ON ? 1 : 0, config.pulse));
+            return;
+        }
+        logger.info("Can not handle command '{}' on channel '{}'", command, channelUID);
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800v3Handler.java:238]: See above

File: bundles/org.openhab.binding.gce/src/main/resources/ESH-INF/thing/channels.xml [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+				<default>false</default>
+			</parameter>
+		</config-description>
+	</channel-type>
+
+	<channel-type id="counter">
+		<item-type>Number</item-type>
+		<label>Counter</label>
+		<config-description>
+			<parameter name="pullFrequency" type="integer" required="false" min="500" max="60000" unit="ms">
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/resources/ESH-INF/thing/channels.xml:167]: Actually it's not a frequency (Hz), but an interval. Sorry, I had to comment on this :)

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+ * @author Seebag - Initial contribution on OH1
+ * @author Gaël L'hopital - Ported and adapted for OH2
+ */
+@NonNullByDefault
+public class Ipx800DeviceConnector extends Thread {
+    private final Logger logger = LoggerFactory.getLogger(Ipx800DeviceConnector.class);
+    private static final int DEFAULT_SOCKET_TIMEOUT = 5000;
+    private static final int DEFAULT_RECONNECT_TIMEOUT = 5000;
+    private static final int MAX_KEEPALIVE_FAILURE = 3;
+    private final static String ENDL = "\r\n";
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java:40]:
```suggestion
    private static final String ENDL = "\r\n";
```

**Review by fwolter — CHANGES_REQUESTED** [review]
There are some formatting issues. You can view them with `mvn spotless:check -Dspotless.check.skip=false` and fix them with `mvn spotless:apply`.

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/action/Ipx800Actions.java [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+            logger.info("Method call reset failed because IPX800 action service ThingHandler is null!");
+        }
+    }
+
+    public static void resetCounter(@Nullable ThingActions actions, int counter) {
+        invokeMethodOf(actions).resetCounter(counter);
+    }
+
+    public static void reset(@Nullable ThingActions actions) {
+        invokeMethodOf(actions).reset();
```
[the remark this answers is not in the document]
Comment (clinique) [reply bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/action/Ipx800Actions.java]: Changed that to use a nullable placeholder argument.

File: bundles/org.openhab.binding.gce/src/main/resources/ESH-INF/thing/ipx800v3.xml [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+<thing:thing-descriptions bindingId="gce" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<thing-type id="ipx800v3">
+		<label>IPX800v3</label>
+		<description>The GCE IPX800v3 device</description>
+
+		<channel-groups>
+			<channel-group id="I" typeId="I"/>
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.gce/src/main/resources/ESH-INF/thing/ipx800v3.xml:11]: Only comment I have left is that in general type are lower or camel case. These id's here are also very short. Does that make sense to the user? Especially I an O can be difficult to distinguish from 1 and 0.

Comment (clinique) [reply bundles/org.openhab.binding.gce/src/main/resources/ESH-INF/thing/ipx800v3.xml:11]: I could rethink this. I've been lazy enough to use the PLC coding as channel groups. I could for sure make it more readable :)

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+     *
+     * @throws IOException
+     */
+    private void connect() throws IOException {
+        disconnect();
+        logger.debug("Connecting {}:{}...", hostname, portNumber);
+        client = new Socket(hostname, portNumber);
+        client.setSoTimeout(DEFAULT_SOCKET_TIMEOUT_MS);
+        client.getInputStream().skip(client.getInputStream().available());
+        in = new BufferedReader(new InputStreamReader(client.getInputStream()));
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java:77]: Can you close the `InputStreamReader`, too?

**Review by fwolter — CHANGES_REQUESTED** [review]
If you complete the readme, I would approve it.

**Comment by clinique** [comment]
> If you complete the readme, I would approve it.

I will, but I'm also currently working on some improvements, and taking in account @Hilbrand comment.

File: bundles/org.openhab.binding.gce/README.md [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+| pullInterval*       | 5000    | No       | Refresh interval (in ms)    |
+
+The binding will query periodically the 'globalstatus.xml' page of the IPX to get fresh informations, this is
+especially usefull for Analog inputs and Counter as modification of these values on PLC side does not trigger any M2M message.
+
+The thing provides four groups of channels.
+
+### Digital Inputs
+
+This represents the inputs of the PLC. Each can be open or closed. They are usually commuted by physical devices like pushbuttons, magnets...
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.gce/README.md:47]: Can you put sentences on a single line. It makes it easier to (future) review diffs.
```suggestion
This represents the inputs of the PLC. Each can be open or closed.
They are usually commuted by physical devices like pushbuttons, magnets...
```

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/model/StatusFileInterpreter.java [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+        try {
+            DocumentBuilder builder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
+            URL url = new URL(String.format(URL_TEMPLATE, hostname));
+
+            URLConnection conn = url.openConnection();
+            Document document = builder.parse(conn.getInputStream());
+            document.getDocumentElement().normalize();
+            doc = document;
+            pushDatas();
+        } catch (IOException | SAXException | ParserConfigurationException e) {
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/model/StatusFileInterpreter.java:71]: You need to close the connection, when you are finished by closing the input stream.

File: bundles/org.openhab.binding.gce/README.md [hunk 24]
```diff
@@ -0,0 +1,162 @@
+# GCE Binding
+
+This binding aims to handle various GCE Electronics equipments.
+IPX800 is a 8 relay webserver from gce-electronics with a lot of possibilities:
+
+* 8 Digital Input
+* 8 Relay (250v/ 10A / channel)
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/README.md:7]:
```suggestion
* 8 Relay (250V / 10A / channel)
```

File: bundles/org.openhab.binding.gce/README.md [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+|--------|-----------------------|--------------------------|-----|-----------------------------------------------------------------------------|
+| analog | `portnumber`          | Number                   |  R  | Value of the port.                                                          |
+| analog | `portnumber`-duration | Number:Time              |  R  | Updated when the port status changes to the duration of the previous state. |
+| analog | `portnumber`-voltage  | Number:ElectricPotential |  R  | Electrical equivalency of the analogic value                                |
+
+#### Configuration
+
+| Property   | Default | Description                                                              |
+|------------|---------|--------------------------------------------------------------------------|
+| histeresis |    0    | If set, the channel will ignore status changes below state + histeresis/2|
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/README.md:124]:
```suggestion
| hysteresis |    0    | If set, the channel will ignore status changes below state + histeresis/2|
```

File: bundles/org.openhab.binding.gce/README.md [hunk 26]
```diff
@@ -0,0 +1,162 @@
...
+|            |         |    or higher than state - histeresis sur 2                               |
```
Comment (fwolter) [inline bundles/org.openhab.binding.gce/README.md:125]: This will render to two rows. Better put in one line.

**Review by fwolter — CHANGES_REQUESTED** [review]
Did  you put the `model` package at the same level as the `internal` package intentionally?

**Comment by clinique** [comment]
> Did you put the `model` package at the same level as the `internal` package intentionally?

No, you're right, this is a mistake.

**Review by fwolter — APPROVED** [review]
LGTM

File: bundles/org.openhab.binding.gce/README.md [hunk 27]
```diff
@@ -0,0 +1,160 @@
...
+
+
+### reset(placeholder)
+
+Restarts the PLC.
+
+* `placeholder` (Integer) - This parameter is not used (can be null).
+
+
+```
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.gce/README.md:157]: It still would be great to have a things and items file example here

File: bundles/org.openhab.binding.gce/src/main/resources/ESH-INF/thing/ipx800v3.xml [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+			<channel-group id="contact" typeId="contact"/>
+			<channel-group id="relay" typeId="relay"/>
+			<channel-group id="counter" typeId="counter"/>
+			<channel-group id="analog" typeId="analog"/>
+		</channel-groups>
+
+		<config-description-ref uri="thing-type:gce:ipx800v3Config"/>
+	</thing-type>
+
+	<channel-group-type id="contact">
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.gce/src/main/resources/ESH-INF/thing/ipx800v3.xml:20]: Maybe use the plural? `contacts` to see the difference between a channel group and channel?

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800v3Handler.java [hunk 29]
```diff
@@ -0,0 +1,393 @@
...
+        }
+
+        connector = new Ipx800DeviceConnector(configuration.hostname, configuration.portNumber, getThing().getUID());
+        parser = new M2MMessageParser(connector, this);
+
+        updateStatus(ThingStatus.UNKNOWN);
+
+        refreshJob = scheduler.scheduleWithFixedDelay(() -> {
+            statusFile.read();
+        }, 3000, configuration.pullInterval, TimeUnit.MILLISECONDS);
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800v3Handler.java:133]:
```suggestion
        refreshJob = scheduler.scheduleWithFixedDelay(statusFile::read, 3000, configuration.pullInterval, TimeUnit.MILLISECONDS);
```

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/model/StatusFileInterpreter.java [hunk 30]
```diff
@@ -0,0 +1,126 @@
...
+        this.hostname = hostname;
+        this.listener = listener;
+    }
+
+    public void read() {
+        try {
+            DocumentBuilder builder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
+            URL url = new URL(String.format(URL_TEMPLATE, hostname));
+
+            URLConnection conn = url.openConnection();
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/model/StatusFileInterpreter.java:66]: It's preferable to use the httpclient. Maybe use HttpUtil here: `HttpUtil.executeUrl("GET", url, 5000);`?

**Review by Hilbrand — CHANGES_REQUESTED** [review]
Just some minor comments and we could merge this binding.

**Review by Hilbrand — APPROVED** [review]
Thanks. LGTM

File: bundles/org.openhab.binding.gce/README.md [hunk 31]
```diff
@@ -0,0 +1,197 @@
...
+Group gIPXOutputs "Outputs" <output>          
+Switch output3 "Chaudière" <furnace> (gIPXOutputs) {channel="gce:ipx800v3:ipx:relay#3"}
+Switch output4 "Lumière Porche" <light> (gIPXOutputs) {channel="gce:ipx800v3:ipx:relay#4"}
+
+```
+
+```
+```
+
+
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.gce/README.md:197]:
```suggestion
```

**Pull request merged**